portality/models/account.py

```diff
@@ -152,12 +149,19 @@ def prep(self):
     @property
     def api_key(self):
         if self.has_role('api'):
-            return self.data.get('api_key', None)
+            # Return the stored api key if it's there, generate one if not (there should be one present if the role is)
```

Comment:
Hmmm. Generating a key if there isn't one is cool .. but it won't really work until the object is .save()-ed. I don't think we should .save() the object on generate_api_key() either.

Is it really necessary to generate one if there isn't one *on read access*? Just return None. It's the caller's responsibility to then .generate_api_key() and .save() . My concern is that this way, say I'm debugging somebody whose key got deleted by an admin. I go into the console and I type account.api_key .. and I get a key! Except it won't work and it will change every time I refresh the object, since it's not persisted. This will also allow you to do ```self.api_key``` on line 128 instead of ```self.data.get('api_key', None)``` - that's already a code smell which tells you the getter isn't really a getter as you're not able to use it as such.

So it'll change the last line of the new test you added above - after deleting an api key, the key should in fact be None :).

Unless I'm missing some kind of use case here where generating the key on read access it required? I'd really look for a way to call .generate_api_key() explicitly in the view function (or whatever the use case that is using the Account model is) instead.